File: EPI/12.12/swift/12.12.swift
public func findDuplicateAndMissing(_ numbers: [Int]) -> (duplicate: Int, missing: Int) {
    let range = 0..<numbers.count

    let numbersXor = numbers.reduce(0, ^)
    let rangeXor = range.reduce(0, ^)

    let duplicateAndMissingXor = numbersXor ^ rangeXor

    let differingBit = duplicateAndMissingXor & ~(duplicateAndMissingXor - 1)

    let filteredNumbersXor = numbers.lazy.filter { $0 & differingBit != 0 }.reduce(0, ^)
    let filteredRangeXor = range.lazy.filter { $0 & differingBit != 0 }.reduce(0, ^)

    let duplicateOrMissing = filteredNumbersXor ^ filteredRangeXor
    let other = duplicateOrMissing ^ duplicateAndMissingXor

    return numbers.contains(duplicateOrMissing) ?
        (duplicate: duplicateOrMissing, missing: other) :
        (duplicate: other, missing: duplicateOrMissing)
}

File: EPI/12.12/typescript/12.12.ts
export default function findDuplicateAndMissing(values: number[]): {duplicate: number, missing: number} {
    const expectedXor = values.reduce((acc, _, i) => acc ^ i, 0);
    const actualXor = values.reduce((acc, v) => acc ^ v, 0);
    const duplicateXorMissing = expectedXor ^ actualXor;

    // Find a bit that is set differently in duplicate and missing.
    // The trick below returns the lowest bit that is set to 1.
    const differingBit = duplicateXorMissing & ~(duplicateXorMissing - 1);

    const expectedFilteredXor = values.reduce((acc, _, i) => (i & differingBit) != 0 ? acc ^ i : acc, 0);
    const actualFilteredXor = values.reduce((acc, v) => (v & differingBit) != 0 ? acc ^ v : acc, 0);
    const eitherDuplicateOrMissing = expectedFilteredXor ^ actualFilteredXor;
    const other = eitherDuplicateOrMissing ^ duplicateXorMissing;

    return values.includes(eitherDuplicateOrMissing) ?
        {duplicate: eitherDuplicateOrMissing, missing: other} :
        {duplicate: other, missing: eitherDuplicateOrMissing};
}
